web/server.go

```diff
@@ -0,0 +1,214 @@
+//     Version: 0.0.1
+//     License: AGPL-3.0 https://opensource.org/licenses/agpl-3.0
+//     Contact: Bruno Michel <bruno@cozycloud.cc> https://cozy.io/
+//
+//     Consumes:
+//     - application/json
+//
+//     Produces:
+//     - application/json
+//
+// swagger:meta
+package web
+
+import (
+	"encoding/json"
+	"html/template"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/cozy/cozy-stack/config"
+	"github.com/cozy/cozy-stack/couchdb"
+	"github.com/cozy/cozy-stack/instance"
+	"github.com/cozy/cozy-stack/web/jsonapi"
+	_ "github.com/cozy/cozy-stack/web/statik" // Generated file with the packed assets
+	"github.com/labstack/echo"
+	"github.com/rakyll/statik/fs"
+)
+
+var templatesList = []string{
+	"login.html",
+}
+
+// Config represents the configuration
+type Config struct {
+	Router    *echo.Echo
+	Assets    string
+	ServeApps func(c echo.Context, domain, slug string) error
+}
+
+type renderer struct {
+	t *template.Template
+	h http.Handler
+}
+
+func (r *renderer) Render(w io.Writer, name string, data interface{}, c echo.Context) error {
+	return r.t.ExecuteTemplate(w, name, data)
+}
+
+func (r *renderer) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	r.h.ServeHTTP(res, req)
+}
+
+func createRenderer(conf *Config) (*renderer, error) {
+	// By default, use the assets packed in the binary
+	if conf.Assets != "" {
+		list := make([]string, len(templatesList))
+		for i, name := range templatesList {
+			list[i] = path.Join(conf.Assets, "templates", name)
+		}
+		t, err := template.ParseFiles(list...)
+		if err != nil {
+			return nil, err
+		}
+		h := http.FileServer(http.Dir(conf.Assets))
+		r := &renderer{t, h}
+		return r, nil
+	}
+
+	statikFS, err := fs.New()
+	if err != nil {
+		return nil, err
+	}
+
+	var t, tmpl *template.Template
+	for _, name := range templatesList {
+		if t == nil {
+			t = template.New(name)
+			tmpl = t
+		} else {
+			tmpl = t.New(name)
+		}
+		f, err := statikFS.Open("/templates/" + name)
+		if err != nil {
+			return nil, err
+		}
+		b, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tmpl.Parse(string(b))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	h := http.FileServer(statikFS)
+	r := &renderer{t, h}
+	return r, nil
+}
+
+func splitHost(host string) (instanceHost string, appSlug string) {
+	parts := strings.SplitN(host, ".", 2)
+	if len(parts) == 2 {
+		return parts[1], parts[0]
+	}
+	return parts[0], ""
+}
+
+// Create returns a new web server that will handle that apps routing given the
+// host of the request. It also adds that the asset handler in /assets/ as well
+// as a template rendering to use c.Render.
+func Create(conf *Config) (*echo.Echo, error) {
+	appsRouter := echo.New()
+	apisRouter := conf.Router
+
+	r, err := createRenderer(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	apisRouter.GET("/assets/*", echo.WrapHandler(http.StripPrefix("/assets/", r)))
+	apisRouter.Renderer = r
```

Comment:
I don't see the purpose of this line? Is it something like a default renderer? But shouldn't the default to be a 404 error rendered by the HTTPErrorHandler?